src/model/CoilCoolingWater.cpp

```diff
@@ -273,48 +273,68 @@ namespace model {
+                      << " cannot be connected directly to an AirLoopHVAC when it's part of a parent CoilSystemCoolingWaterHeatExchangerAssisted. "
+                         "Please call CoilSystemCoolingWaterHeatExchangerAssisted::addToNode instead");
+          return false;
+        }
 
-      } else {
+        if (t_containingHVACComponent->optionalCast<CoilSystemCoolingWater>()
+            && t_containingHVACComponent->cast<CoilSystemCoolingWater>().coolingCoil().handle() == this->handle()) {
+          LOG(Warn, this->briefDescription()
+                      << " cannot be connected directly to an AirLoopHVAC when it's the primary cooling coil of a parent CoilSystemCoolingWater. "
+                         "Please call CoilSystemCoolingWater::addToNode instead");
+          return false;
+        }
+      }
 
-        success = WaterToAirComponent_Impl::addToNode(node);
-        auto t_containingZoneHVACComponent = containingZoneHVACComponent();
+      success = WaterToAirComponent_Impl::addToNode(node);
+      if (!success) {
+        return false;
+      }
+      // We only add the controller if we are not part of a containing ZoneHVACComponent, and we have a PlantLoop assigned
+      if (containingZoneHVACComponent() || !waterInletModelObject()) {
+        // We're done
+        return true;
+      }
 
-        if (success && (!t_containingZoneHVACComponent)) {
-          if (auto t_waterInletModelObject = waterInletModelObject()) {
-            if (auto oldController = controllerWaterCoil()) {
-              if (!openstudio::istringEqual(oldController->action().get(), "Reverse")) {
-                LOG(Warn,
-                    briefDescription()
-                      << " has an existing ControllerWaterCoil with action set to something else than 'Reverse'. Make sure this is what you want");
-              }
-            } else {
-              if (t_containingHVACComponent && t_containingHVACComponent->optionalCast<CoilSystemCoolingWater>()) {
-                // no-op
-              } else {
-                Model t_model = model();
-                ControllerWaterCoil controller(t_model);
-                controller.getImpl<ControllerWaterCoil_Impl>()->setWaterCoil(getObject<HVACComponent>());
-                controller.setAction("Reverse");
-              }
+      // We don't need a controller if we are part of a CoilSystemCoolingWater (directly, or indirectly via a
+      // CoilSystemCoolingWaterHeatExchangerAssisted)
+      bool needAController = true;
+      if (t_containingHVACComponent) {
+        if (t_containingHVACComponent->optionalCast<CoilSystemCoolingWater>()) {
+          needAController = false;
+        } else if (auto hxAssisted_ = t_containingHVACComponent->optionalCast<CoilSystemCoolingWaterHeatExchangerAssisted>()) {
+          if (auto parent_ = hxAssisted_->containingHVACComponent()) {
+            if (parent_->optionalCast<CoilSystemCoolingWater>()) {
+              needAController = false;
```

Comment:
We need a ControllerWaterCoil except if we're directly or indirectly (via a CoilSystemCoolingWaterHeatExchangerAssisted) part of a CoilSystemCoolingWater.